sky/exceptions.py

```diff
@@ -29,6 +29,18 @@ def with_failover_history(self, failover_history: List[Exception]) -> None:
         return self
 
 
+class SpotJobFailBeforeProvisionError(Exception):
```

Comment:
Curious what's the reason for adding this? I'm not sure if this is clearer than before. E.g., "before provision" may be ambiguous, because we may have encountered invalid name on one cloud, and resource unavailable on another. Does that count as "before provision"? The previous impl seems good enough.

Reply:
The case mentioned above should not happen when this error is raised. Here are the reasons I added this error:
1. I found the previous implementation hard to distinguish the exception raised after maximum retry and the exception raised because the `ResourcesUnavailableError` has an empty `failover_history`. I think it would be better to explicitly distinguish the error.
2. We will only raise the exception iff the optimizer cannot find the feasible resources, or none of the failover is because of resources unavailability. (https://github.com/skypilot-org/skypilot/blob/1bf79933e577ec7bd667874037a0b4f5c30541ee/sky/spot/recovery_strategy.py#L233-L263)
3. It would be quite verbose to check whether the `failover_history` contains `ResourcesUnavailableError` in both `recovery_strategy.py` and `controller.py` in the previous implementation.